(dyn ArrayFormatterFactory + 'static) can't be safely shared between threads

**Describe the bug**

When I tried the new version of arrow with DataFusion in this PR:
- https://github.com/apache/datafusion/pull/18820

Compilation fails like this:

```
 Compiling datafusion-sql v51.0.0 (/Users/andrewlamb/Software/datafusion/datafusion/sql)
error[E0277]: `(dyn ArrayFormatterFactory + 'static)` cannot be shared between threads safely
   --> datafusion/physical-expr/src/expressions/cast.rs:130:23
    |
130 | impl PhysicalExpr for CastExpr {
    |                       ^^^^^^^^ `(dyn ArrayFormatterFactory + 'static)` cannot be shared between threads safely
    |
    = help: within `CastExpr`, the trait `Sync` is not implemented for `(dyn ArrayFormatterFactory + 'static)`
    = note: required because it appears within the type `&'static (dyn ArrayFormatterFactory + 'static)`
note: required because it appears within the type `Option<&'static (dyn ArrayFormatterFactory + 'static)>`
   --> /Users/andrewlamb/.rustup/toolchains/1.91.0-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/option.rs:594:10
    |
594 | pub enum Option<T> {
    |          ^^^^^^
note: required because it appears within the type `arrow::util::display::FormatOptions<'static>`
   --> /Users/andrewlamb/.cargo/git/checkouts/arrow-rs-a228b2f48bc25539/f7d41c8/arrow-cast/src/display.rs:63:12
    |
 63 | pub struct FormatOptions<'a> {
    |            ^^^^^^^^^^^^^
note: required because it appears within the type `CastOptions<'static>`
   --> /Users/andrewlamb/.cargo/git/checkouts/arrow-rs-a228b2f48bc25539/f7d41c8/arrow-cast/src/cast/mod.rs:76:12
    |
 76 | pub struct CastOptions<'a> {
    |            ^^^^^^^^^^^
note: required because it appears within the type `CastExpr`
   --> datafusion/physical-expr/src/expressions/cast.rs:46:12
```

**To Reproduce**
See https://github.com/apache/datafusion/pull/18820


**Expected behavior**
I should be able to keep on compiling what previously worked
**Additional context**
<!--
Add any other context about the problem here.
-->

--- a/arrow-cast/src/display.rs
+++ b/arrow-cast/src/display.rs
@@ -352,7 +352,7 @@ impl<'a> FormatOptions<'a> {
 ///        &FormatOptions::new().with_formatter_factory(Some(&MyFormatters {}))
 /// );
 /// ```
-pub trait ArrayFormatterFactory: Debug {
+pub trait ArrayFormatterFactory: Debug + Send + Sync {
     /// Creates a new [`ArrayFormatter`] for the given [`Array`] and an optional [`Field`]. If the
     /// default implementation should be used, return [`None`].
     ///
@@ -1338,6 +1338,19 @@ mod tests {
     #[test]
     fn test_const_options() {
         assert_eq!(TEST_CONST_OPTIONS.date_format, Some("foo"));
+    }
+
+    /// See https://github.com/apache/arrow-rs/issues/8875
+    #[test]
+    fn test_options_send_sync() {
+        fn assert_send_sync<T>()
+        where
+            T: Send + Sync,
+        {
+            // nothing – the compiler does the work
+        }
+
+        assert_send_sync::<FormatOptions<'static>>();
     }
 
     #[test]

